### PR TITLE
test: add e2e tests in kong 3.0 compatibility test action

### LINKED ISSUE #2870: Test KIC 2.6 `main` against Kong OSS 3.0.0-alpha.1 and kong-ee nightly
Ensure that the below testing configurations pass:

Kong OSS:
- `make test.integration` (both postgres and dbless) vs Kong OSS 3.0.0-alpha.1
- `make test.e2e` vs Kong OSS 3.0.0-alpha.14

Kong Enterprise:
- `make test.integration.enterprise.postgres` against ~Kong EE nightly~ the internal image labeled `3.0.x.x` (covered by #2850).

Let's report bugs referencing this issue for failures under investigation/fix.

Acceptance criteria
- [ ] The `test.integration` dbless against Kong OSS yields no failures
- [ ] The `test.integration` postgres against Kong OSS yields no failures
- [ ] The `test.integration.enterprise.postgres` against Kong EE yields no failures

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

Add e2e tests into kong 3.0 compatibility tests.

**Which issue this PR fixes**:

fixes #2870 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: .github/workflows/kong3.0_compatibility.yaml [hunk 1]
```diff
@@ -86,8 +88,94 @@ jobs:
...
+      - name: setup kong enterprise license
+        uses: Kong/kong-license@master
+        id: license
+        with:
+          # PULP_PASSWORD secret is set in "Configure ci" environment
+          password: ${{ secrets.PULP_PASSWORD }}
+
+      - name: run enterprise cases of e2e test
+        run: make test.e2e ## TODO: the config runs tests for both non-enterprise and enterprise tests. Should we run enterprise tests only in this step?
+        ## run tests without KONG_LICENSE_DATA env to skip enterprise tests.
```
Comment (pmalek) [inline .github/workflows/kong3.0_compatibility.yaml:163]: I'd separate those so that it's easier to figure out what failed when a failure occurs.

And ideally I'd put those into separate jobs so that they can run in parallel and independently of each other.

Comment (randmonkey) [reply .github/workflows/kong3.0_compatibility.yaml:163]: separate what? integration vs e2e or OSS vs enterprise?

Comment (pmalek) [reply .github/workflows/kong3.0_compatibility.yaml:163]: This runs inside a job called `e2e-tests` (line 95) which runs both OS and enterprise.

I suggest to split those into 2 so that they run in parallel, separated. WDYT?

File: .github/workflows/kong3.0_compatibility.yaml [hunk 2]
```diff
@@ -86,8 +88,94 @@ jobs:
...
+          TEST_KONG_PULL_PASSWORD: ${{ secrets.GHA_KONG_ORG_DOCKERHUB_PUBLIC_TOKEN }}
+          KONG_CLUSTER_VERSION: "v1.24.2"
+          ISTIO_VERSION: "v1.14.1"
+          NCPU: 1
+
+      - name: upload diagnostics for e2e enterprise 
+        if: ${{ always() }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: diagnostics-e2e-tests-enterpeise
```
Comment (rainest) [inline .github/workflows/kong3.0_compatibility.yaml:179]:
```suggestion
          name: diagnostics-e2e-tests-enterprise
```

File: .github/workflows/kong3.0_compatibility.yaml [hunk 3]
```diff
@@ -86,8 +88,94 @@ jobs:
...
+
+      - name: setup kong enterprise license
+        uses: Kong/kong-license@master
+        id: license
+        with:
+          # PULP_PASSWORD secret is set in "Configure ci" environment
+          password: ${{ secrets.PULP_PASSWORD }}
+
+      - name: run enterprise cases of e2e test
+        run: make test.e2e ## TODO: the config runs tests for both non-enterprise and enterprise tests. Should we run enterprise tests only in this step?
```
Comment (rainest) [inline .github/workflows/kong3.0_compatibility.yaml:162]: Enterprise breaking shared functionality that OSS tests as well isn't unheard of, although it's much less common than it was in the past (merging the workspace code into OSS removed a lot of the different behavior), so I'd say run them all for both. It's a cronjob test, it doesn't really need to be particularly time-efficient.

File: .github/workflows/kong3.0_compatibility.yaml (hunk 1 above)
Comment (randmonkey) [reply .github/workflows/kong3.0_compatibility.yaml:163]: Running them in parallel would cost less time to get the result, so +1 for this.

File: .github/workflows/kong3.0_compatibility.yaml [hunk 4]
```diff
@@ -86,8 +88,103 @@ jobs:
...
+        # PULP_PASSWORD secret is set in "Configure ci" environment
+        password: ${{ secrets.PULP_PASSWORD }}
+
+    - name: run enterprise cases of e2e test
+      run: make test.e2e
+      env:
+        TEST_KONG_CONTROLLER_IMAGE_LOAD: kong/kubernetes-ingress-controller:ci
+        TEST_KONG_CONTROLLER_IMAGE_OVERRIDE: kong/kubernetes-ingress-controller:ci
+        TEST_KONG_IMAGE_LOAD: kong/kong-gateway-internal:3.0.x.x-alpine
+        KONG_LICENSE_DATA: ${{ steps.license.outputs.license }}
```
Comment (pmalek) [inline .github/workflows/kong3.0_compatibility.yaml:177]: Looking at the skipping logic we may want to change how tests are skipped in e2e tests suite, i.e. use a separate env var to signal whether to run enterprise tests or not.

Or use a convention where each test enterprise test would have the prefix `TestE2EEnterprise...` and each OS `TestE2EOS....` or something similar.

We can do this in a separate PR obviously 👍

File: .github/workflows/kong3.0_compatibility.yaml [hunk 5]
```diff
@@ -86,8 +88,103 @@ jobs:
...
+        TEST_KONG_PULL_USERNAME: ${{ secrets.GHA_DOCKERHUB_PULL_USER }}
+        TEST_KONG_PULL_PASSWORD: ${{ secrets.GHA_KONG_ORG_DOCKERHUB_PUBLIC_TOKEN }}
+        KONG_CLUSTER_VERSION: "v1.24.2"
+        ISTIO_VERSION: "v1.14.1"
+        NCPU: 1
+
+    - name: upload diagnostics for e2e enterprise 
+      if: ${{ always() }}
```
Comment (pmalek) [inline .github/workflows/kong3.0_compatibility.yaml:185]: Why not when the test fail?

**Review by pmalek — APPROVED** [review]

File: .github/workflows/kong3.0_compatibility.yaml (hunk 5 above)
Comment (randmonkey) [reply .github/workflows/kong3.0_compatibility.yaml:185]: when test passed, their would be no such files and actually no files will be updated.

File: .github/workflows/kong3.0_compatibility.yaml (hunk 4 above)
Comment (randmonkey) [reply .github/workflows/kong3.0_compatibility.yaml:177]: Yep, would you create a new issue for that?

Comment (pmalek) [reply .github/workflows/kong3.0_compatibility.yaml:177]: Ok done, #2912
